### ISSUE TITLE
ComputeError: RecursionError during read_parquet

### ISSUE THREAD

**Issue by CHDev93** [pola-rs/polars#12589]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
pl.read_parquet("foo.parquet")
```

### Log output

```shell
In [3]: df = pl.read_parquet(path)
---------------------------------------------------------------------------
ComputeError                              Traceback (most recent call last)
Cell In[3], line 1
----> 1 df = pl.read_parquet(path)

File ~/env/lib/python3.10/site-packages/polars/io/parquet/functions.py:131, in read_parquet(source, columns, n_rows, use_pyarrow, memory_map, storage_options, parallel, row_count_name, row_count_offset, low_memory, pyarrow_options, use_statistics, rechunk)
    120     import pyarrow.parquet
    122     return from_arrow(  # type: ignore[return-value]
    123         pa.parquet.read_table(
    124             source_prep,
   (...)
    128         )
    129     )
--> 131 return pl.DataFrame._read_parquet(
    132     source_prep,
    133     columns=columns,
    134     n_rows=n_rows,
    135     parallel=parallel,
    136     row_count_name=row_count_name,
    137     row_count_offset=row_count_offset,
    138     low_memory=low_memory,
    139     use_statistics=use_statistics,
    140     rechunk=rechunk,
    141 )

File ~/env/lib/python3.10/site-packages/polars/dataframe/frame.py:851, in DataFrame._read_parquet(cls, source, columns, n_rows, parallel, row_count_name, row_count_offset, low_memory, use_statistics, rechunk)
    840 scan = scan_parquet(
    841     source,
    842     n_rows=n_rows,
   (...)
    847     low_memory=low_memory,
    848 )
    850 if columns is None:
--> 851     return scan.collect()
    852 elif is_str_sequence(columns, allow_str=False):
    853     return scan.select(columns).collect()

File ~/env/lib/python3.10/site-packages/polars/utils/deprecation.py:100, in deprecate_renamed_parameter.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
     95 @wraps(function)
     96 def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
     97     _rename_keyword_argument(
     98         old_name, new_name, kwargs, function.__name__, version
     99     )
--> 100     return function(*args, **kwargs)

File ~/env/lib/python3.10/site-packages/polars/lazyframe/frame.py:1788, in LazyFrame.collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, slice_pushdown, comm_subplan_elim, comm_subexpr_elim, no_optimization, streaming, _eager)
   1775     comm_subplan_elim = False
   1777 ldf = self._ldf.optimization_toggle(
   1778     type_coercion,
   1779     predicate_pushdown,
   (...)
   1786     _eager,
   1787 )
-> 1788 return wrap_df(ldf.collect())

ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: ComputeError: RecursionError: maximum recursion depth exceeded while calling a Python object
```

### Issue description

Reading certain parquet files leads to an infinite recursion. Setting `use_pyarrow=True` resolves the issue. I _think_ the issue is due to reading this file from a fileshare. I'm not familiar enough with the inner workings of fileshares to have a good handle on what is different that could cause this. 

My observations thus far are 
- it doesn't happen to all files, but for files which it does occur, it occurs reliably
- sometimes I get errors about the parquet file being malformed (unable to find magic bytes)
- using the pyarrow backend always solves the issue

### Expected behavior

Reads the file similarly to pyarrow

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.14
Index type:          UInt32
Platform:            Linux-5.15.86-flatcar-x86_64-with-glibc2.35
Python:              3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         3.0.0
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              2023.9.2
gevent:              <not installed>
matplotlib:          3.8.0
numpy:               1.23.5
openpyxl:            <not installed>
pandas:              1.5.3
pyarrow:             11.0.0
pydantic:            2.4.2
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>```

</details>

**Comment by ritchie46** [comment]
Can you share the file?

**Comment by CHDev93** [comment]
Bunch of company restrictions means I can't share the file :/ But the schema is below if that helps at all

```python
pl.from_pandas(pd.read_parquet(path)).schema
Out[27]:
OrderedDict([('col1', Int64),
             ('col2', Int64),
             ('col3', Date),
             ('col4', Int32),
             ('col5', Int32),
             ('col5', Int32),
             ('col6', Int32)])
```
Have seen it happen on various shaped files. I've also confirmed that by copying the file locally (from the remote file system), the problem goes away. 

Appreciate it's probably not possible to repro this one on your end but was hoping that the RecursionDepth exceeded error might give some hint as to where the read is going wrong. Seems like some assumption about the filesystem doesn't hold for network file systems and triggers this code path.

If nothing else, just wanted to open this to document the issue in case anyone else comes across it.

**Comment by cmdlineluser** [comment]
Can you give more details about the fileshare / `path` maybe? (s3, gcp, azure?)

The `ComputeError: ComputeError:  ComputeError` looks like there is some weird recursion going on with the exceptions themselves.

**Comment by deanm0000** [comment]
Can you create a new parquet, generated with random data but is saved in the same manner as the problem ones such that it, too, can't be opened and isn't proprietary?

**Comment by CHDev93** [comment]
@cmdlineluser It's an internal distributed file system using either the SMB or NFS protocol

**Comment by deanm0000** [comment]
I see, didn't realize it worked locally. Sharing an example file can't hurt incase somebody has a convenient means of replicating the file share but I don't know for sure. Can you pinpoint which it is between SMB and NFS? 

Is it file size dependent? Are bigger files more likely to fail than smaller ones? 

Assuming that *is* the case and without really knowing much about anything, I'm guessing maybe small files work because it gets all the data in a single chunk but that something weird happens if it needs to transfer multiple chunks of a file.

**Comment by CHDev93** [comment]
This particular file was on an NFS share. Have observed the error on files as small as (910,7) but also on larger files as well so it's not obviously file size dependent.

**Comment by CrispyCrafter** [comment]
I have managed to replicate this bug as follows:

```
import polars as pl

path_with_special_chars = 'SOME_PATH/[SOME_PARTITION_NAME]/file.parquet'
clean_path = 'SOME_PATH/SOME_PARTITION_NAME/file.parquet'

pl.read_parquet(path_with_special_chars) # Recursion Error
pl.read_parquet(clean_path) # Processes normally
```

It seems that the file path is being parsed in such a way that it is trying to recursively resolve partitions ? or some other aspect if the path contains certain special characters. This observation echoes the previous findings where once the file was downloaded locally it seemingly work without error.

Could this be related to some quirk of fsspec or some other internal file path resolution?

**Comment by cmdlineluser** [comment]
@CrispyCrafter I can reproduce that.

```python
import polars as pl
import tempfile
from   pathlib import Path

with tempfile.TemporaryDirectory() as tmpdir:
    hive_dir = Path(tmpdir) / "[group]"
    Path(hive_dir).mkdir()

    filename = str(hive_dir / "file.parquet")
    pl.DataFrame({"foo": [1]}).write_parquet(filename)

    df = pl.read_parquet(filename)
    # ComputeError: RecursionError: maximum recursion depth exceeded in comparison
```

It also happens with `scan_parquet` when `.collect()` is called.

Using `[group` instead I get `ComputeError: invalid glob pattern given` so it seems to have something to do with `[...]` being parsed as a glob.

Other glob chars `?` and `*` do not cause the error.

**Comment by CrispyCrafter** [comment]
I discovered in a package that integrates polars as an IO layer
https://github.com/dagster-io/dagster/issues/21451

**Comment by ritchie46** [comment]
fixed by #15927 and #15928
